// @storybook/angular/dist/preset.js
import CJS_COMPAT_NODE_URL_06if1gnmlzgm from 'node:url';
import CJS_COMPAT_NODE_PATH_06if1gnmlzgm from 'node:path';
import CJS_COMPAT_NODE_MODULE_06if1gnmlzgm from "node:module";

var __filename = CJS_COMPAT_NODE_URL_06if1gnmlzgm.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_06if1gnmlzgm.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_06if1gnmlzgm.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------

// src/preset.ts
import { fileURLToPath } from "node:url";
var addons = [
  fileURLToPath(import.meta.resolve("@storybook/angular/server/framework-preset-angular-cli")),
  fileURLToPath(import.meta.resolve("@storybook/angular/server/framework-preset-angular-ivy"))
], previewAnnotations = async (entries = [], options) => {
  let config = fileURLToPath(import.meta.resolve("@storybook/angular/client/config")), annotations = [...entries, config];
  if (options.enableProdMode) {
    let previewProdPath = fileURLToPath(
      import.meta.resolve("@storybook/angular/client/preview-prod")
    );
    annotations.unshift(previewProdPath);
  }
  let docsConfig = await options.presets.apply("docs", {}, options);
  if (Object.keys(docsConfig).length > 0) {
    let docsConfigPath = fileURLToPath(
      import.meta.resolve("@storybook/angular/client/docs/config")
    );
    annotations.push(docsConfigPath);
  }
  return annotations;
}, core = async (config, options) => {
  let framework = await options.presets.apply("framework");
  return {
    ...config,
    builder: {
      name: import.meta.resolve("@storybook/builder-webpack5"),
      options: typeof framework == "string" ? {} : framework.options.builder || {}
    }
  };
}, typescript = async (config) => ({
  ...config,
  skipCompiler: !0
});
export {
  addons,
  core,
  previewAnnotations,
  typescript
};
